dimod/discrete/discrete_quadratic_model.py

```diff
@@ -383,5 +521,210 @@ def set_quadratic_case(self, u, u_case, v, v_case, bias):
+                     num_cases=dqm.num_cases(),
+                     num_case_interactions=dqm.num_case_interactions(),
+                     num_variable_interactions=dqm.num_variable_interactions(),
+                     variables=not (ignore_labels or dqm.variables.is_range),
+                     )
+
+            it is padded with spaces to make the entire length of the header
+            divisible by 64.
+
+            **DQM Data**
+
+            The first 4 bytes are exactly `"BIAS"`
+
+            The next 4 bytes form a little-endian unsigned int, the length of
+            the DQM data `DATA_LEN`.
+
+            The next `DATA_LEN` bytes are the vectors as returned by
+            :meth:`DiscreteQuadraticModel.to_numpy_vectors` saved using
+            :func:`numpy.save`.
+
+            **Variable Data**
+
+            The first 4 bytes are exactly "VARS".
+
+            The next 4 bytes form a little-endian unsigned int, the length of
+            the variables array `VARIABLES_LENGTH`.
+
+            The next VARIABLES_LENGTH bytes are a json-serialized array. As
+            constructed by `json.dumps(list(bqm.variables)).
+
+        .. _NPY format: https://docs.scipy.org/doc/numpy/reference/generated/numpy.lib.format.html
+
+        See Also:
+            :meth:`DiscreteQuadraticModel.from_file`
+
+        """
+
+        file = tempfile.SpooledTemporaryFile(max_size=max_size)
+
+        # attach the header
+        header_parts = [DQM_MAGIC_PREFIX,
+                        VERSION,
+                        bytes(4),  # placeholder for HEADER_LEN
+                        ]
+
+        index_labeled = ignore_labels or self.variables.is_range
+
+        header_data = json.dumps(
+            dict(num_variables=self.num_variables(),
+                 num_cases=self.num_cases(),
+                 num_case_interactions=self.num_case_interactions(),
+                 num_variable_interactions=self.num_variable_interactions(),
+                 variables=not index_labeled,
+                 ),
+            sort_keys=True).encode('ascii')
+
+        header_parts.append(header_data)
+
+        # make the entire header length divisible by 64
+        length = sum(len(part) for part in header_parts)
+        if length % 64:
+            padding = b' '*(64 - length % 64)
+        else:
+            padding = b''
+        header_parts.append(padding)
+
+        HEADER_LEN = len(padding) + len(header_data)
+        header_parts[2] = np.dtype('<u4').type(HEADER_LEN).tobytes()
+
+        for part in header_parts:
+            file.write(part)
+
+        # the section containing most of the data, encoded with numpy
+        self._to_file_numpy(file, compressed)
+
+        if not index_labeled:
+            file.write(VariablesSection(self.variables).dumps())
+
+        file.seek(0)
+
+        return file
+
+    def to_numpy_vectors(self, return_labels=False):
+        """Convert the DQM to five numpy vectors.
```

Comment:
Just a thought: maybe it's time to start returning `namedtuple`.

Reply:
Yeah, I was trying to maintain similarity to `BQM.to_numpy_vectors`, but probably practicality trumps similarity. `namedtuple` seems cleaner. Maybe I'll add an option to return a "flat" `namedtuple` from the BQM ones as well as an argument and put up a deprecation notice.

Reply:
No need, `namedtuple` is `tuple`'s subclass, so existing code would not break. At least destructuring will work. Doubt anyone checks exact rtype.

Reply:
I was more referring to the fact that I returned a nested tuple `(case_starts, linear_biases, (irow, icol, qdata))`. In either case, I can just return a nested namedtuple as well, which is what I will do. I think if I was writing it from scratch I would just return a flat one but :shrug:

Reply:
Oh, I see. Well, we'll fix it in Ocean 5. :smile: